lib/storage/src/chunks/getChunkStream.ts

```diff
@@ -0,0 +1,39 @@
+import { RawDataPart } from "../Upload";
+import { Buffer } from "buffer";
+
+interface Buffers {
+  chunks: Buffer[];
+  length: number;
+}
+
+export async function* getChunk<T>(
+  data: T,
+  partSize: number,
+  getNextData: (data: T) => AsyncGenerator<Buffer>
+): AsyncGenerator<RawDataPart> {
+  let partNumber = 1;
+  const currentBuffer: Buffers = { chunks: [], length: 0 };
```

Comment:
i think you could keep only one buffer and simplify a bit the code, eg:

```
  let partNumber = 1;
  let currentBuffer = Buffer.of();

  for await (const datum of getNextData(data)) {
    currentBuffer = Buffer.concat([currentBuffer, datum]);

    while (currentBuffer.length >= partSize) {
      yield {
        partNumber,
        data: currentBuffer.slice(0, partSize),
      };

      //update the buffer
      currentBuffer = currentBuffer.slice(partSize);
      partNumber += 1;
    }
  }
  yield {
    partNumber,
    data: currentBuffer,
  };
```

Reply:
See https://github.com/aws/aws-sdk-js-v3/pull/1955 why this is a bad idea. Buffer.concat is surprisingly slow and should be avoided at all cost. Doing it in a loop is even worse as you copy the buffers over and over as they grow larger. Also increasing the memory footprint of the node process as it might not get a chance to GC.

Reply:
Yep, building off that, thats why this tries to wait until the last possible moment to run the concat.

Reply:
you're right, i miss that point!